Peer-graded Assignment: Build a Regression Model in Keras Part C

Created by Kuan Yew Cheng, Date: 23.10.2020

In [1]:
# import important libraries

import pandas as pd
import numpy as np

Generate the dataframe from dataset concrete_data

In [2]:
# download the data and read it into a pandas dataframe
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

As we knew from the lab 3 that the data is cleaned and is ready to be used to build our model.

Split data into predictors and target

In [3]:
# all columns except strength is our predictors column and target is our strength column
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
target = concrete_data['Strength']

Next, normalize our data before applying to the neural network model

In [4]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

# number of predictors
n_cols = predictors_norm.shape[1] 

Build our neural network model

In [5]:
# import Keras library and other related libraries
import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import mean_squared_error

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

First, define the model

In [6]:
# define regression model
def regression_model():
    #create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Split our dataset into train and test set with normalized dataset

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


Build the model and fit our train data in

In [8]:
# build the model
model = regression_model()

Error Evaluation and create a list of mse

In [9]:
mean_squared_errors = []

for i in range(50):   
    # fit the model
    model.fit(X_train, y_train, epochs=100, verbose=2)

    # Evaluate the model on the test data
    y_hat = model.predict(X_test)
    mse = mean_squared_error(y_test, y_hat)
    mean_squared_errors.append(mse)

print(mean_squared_errors)

Epoch 1/100
 - 1s - loss: 1543.4726
Epoch 2/100
 - 0s - loss: 1528.3880
Epoch 3/100
 - 0s - loss: 1512.9414
Epoch 4/100
 - 0s - loss: 1496.6439
Epoch 5/100
 - 0s - loss: 1479.5945
Epoch 6/100
 - 0s - loss: 1461.6235
Epoch 7/100
 - 0s - loss: 1442.4121
Epoch 8/100
 - 0s - loss: 1422.2030
Epoch 9/100
 - 0s - loss: 1400.5944
Epoch 10/100
 - 0s - loss: 1377.8676
Epoch 11/100
 - 0s - loss: 1353.7866
Epoch 12/100
 - 0s - loss: 1328.3723
Epoch 13/100
 - 0s - loss: 1301.6347
Epoch 14/100
 - 0s - loss: 1273.8528
Epoch 15/100
 - 0s - loss: 1244.8469
Epoch 16/100
 - 0s - loss: 1215.0671
Epoch 17/100
 - 0s - loss: 1183.6995
Epoch 18/100
 - 0s - loss: 1151.5305
Epoch 19/100
 - 0s - loss: 1118.7977
Epoch 20/100
 - 0s - loss: 1085.1208
Epoch 21/100
 - 0s - loss: 1050.6909
Epoch 22/100
 - 0s - loss: 1016.1042
Epoch 23/100
 - 0s - loss: 981.3433
Epoch 24/100
 - 0s - loss: 946.2817
Epoch 25/100
 - 0s - loss: 910.8587
Epoch 26/100
 - 0s - loss: 875.9694
Epoch 27/100
 - 0s - loss: 841.2639
Epoch 28/100
 -

In [12]:
#mean
np.mean(mean_squared_errors)

print("Mean after 100 epochs: %3.3f"  %(np.mean(mean_squared_errors)))

Mean after 100 epochs: 46.136


In [13]:
# standatd deviation
np.std(mean_squared_errors)
print("Std after 100 epochs: %3.3f"  %(np.std(mean_squared_errors)))

Std after 100 epochs: 18.591


The mean of mean squared errors of step C decreased to 46.14 from 57.23 compared to step B after using 100 epochs for training instead of 50 epochs.